src/modules/musig/main_impl.h

```diff
@@ -0,0 +1,586 @@
+/**********************************************************************
+ * Copyright (c) 2018 Andrew Poelstra                                 *
+ * Distributed under the MIT software license, see the accompanying   *
+ * file COPYING or http://www.opensource.org/licenses/mit-license.php.*
+ **********************************************************************/
+
+#ifndef _SECP256K1_MODULE_MUSIG_MAIN_
+#define _SECP256K1_MODULE_MUSIG_MAIN_
+
+#include "include/secp256k1.h"
+#include "include/secp256k1_musig.h"
+#include "hash.h"
+
+/* Computes ell = SHA256(pk[0], ..., pk[np-1]) */
+static int secp256k1_musig_compute_ell(const secp256k1_context *ctx, unsigned char *ell, const secp256k1_pubkey *pk, size_t np) {
+    secp256k1_sha256 sha;
+    size_t i;
+
+    secp256k1_sha256_initialize(&sha);
+    for (i = 0; i < np; i++) {
+        unsigned char ser[33];
+        size_t serlen = sizeof(ser);
+        if (!secp256k1_ec_pubkey_serialize(ctx, ser, &serlen, &pk[i], SECP256K1_EC_COMPRESSED)) {
+            return 0;
+        }
+        secp256k1_sha256_write(&sha, ser, serlen);
+    }
+    secp256k1_sha256_finalize(&sha, ell);
+    return 1;
+}
+
+/* Compute r = SHA256(ell, idx) */
+static void secp256k1_musig_coefficient(secp256k1_scalar *r, const unsigned char *ell, size_t idx) {
+    secp256k1_sha256 sha;
+    unsigned char buf[32];
+    secp256k1_sha256_initialize(&sha);
+    secp256k1_sha256_write(&sha, ell, 32);
+    while (idx > 0) {
+        unsigned char c = idx;
+        secp256k1_sha256_write(&sha, &c, 1);
+        idx /= 0x100;
```

Comment:
I somehow feel that a right shift by 8 is much more readable. :D But it's correct.

Reply:
agreed